neo4j/session_with_context.go

```diff
@@ -352,8 +352,11 @@ func (s *sessionWithContext) BeginTransaction(ctx context.Context, configurers .
 			s.pool.Return(ctx, tx.conn)
 			tx.err = errorutil.CombineAllErrors(tx.err, bookmarkErr)
 			tx.conn = nil
+			// TODO not sure if clearing the results here is necessary.
+			tx.results = nil
 			s.explicitTx = nil
 		},
+		results: make([]ResultWithContext, 0),
```

Comment:
disclaimer: I'm not sure this will make the code better

it's likely going to be a bit more lightweight to store a slice of `func(error)`
assuming the change looks like:
```go
resultCallbacks : make([]func(error))
```

the tx code below would become:
```go
	for _, resultCallback := range tx.resultCallbacks {
			// TODO possibly replace err with the a new 'ResultInvalid' error
			resultCallback(err)
		}
```
and
```go
tx.resultCallbacks = append(tx.resultCallbacks, result.onTxError)
```

Reply:
Alternative: have a shared struct like `TransactionState` that is shared between the tx and its results. State could be one of `Open`, `Closed`, `Failed` with some `err`.